### PR TITLE
[RLlib] Example for custom multi-agent vetor env. Bug fixes related to this use-case.

### PR DESCRIPTION
**Pull request by sven1977**
Example for custom multi-agent vetor env. Bug fixes related to this use-case.

* RLlib currently does not support defining a custom VectorEnv, which has n underlying MultiAgentEnvs (this PR fixes the bugs related to this scenario).
* Adds a new example script to illustrate how to write a custom VectorEnv with one or more underlying MultiAgentEnvs.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/env/multi_agent_env.py [hunk 1]
```diff
@@ -408,37 +408,42 @@ def __init__(self, config=None):
             else:
                 self.agents = [env_name_or_creator(config) for _ in range(num)]
             self.dones = set()
-            self.observation_space = self.agents[0].observation_space
-            self.action_space = self.agents[0].action_space
+            self._spaces_in_preferred_format = True
+            self.observation_space = gym.spaces.Dict(
```
Comment (gjoliver) [inline rllib/env/multi_agent_env.py:412]: can I make a quick coding suggestion, as written, the keys of self.observation_space and self.action_space are only correct if self._agent_ids stays set(range(num)).
a safer way to do this maybe to create self._agent_ids first, then:

```
self._agent_ids = set(range(num))
self.observation_space = gym.spaces.Dict(
    {i: a.observation_space for i, a in zip(self._agent_ids, self.agents)}
)
self.action_space = gym.spaces.Dict(
    {i: a.action_space for i, a in zip(self._agent_ids, self.agents)}
)
```
definite a nit.

File: rllib/env/multi_agent_env.py [hunk 2]
```diff
@@ -408,37 +408,42 @@ def __init__(self, config=None):
...
+                {i: a.action_space for i, a in enumerate(self.agents)}
+            )
             self._agent_ids = set(range(num))
 
         @override(MultiAgentEnv)
         def observation_space_sample(self, agent_ids: list = None) -> MultiAgentDict:
             if agent_ids is None:
                 agent_ids = list(range(len(self.agents)))
-            obs = {agent_id: self.observation_space.sample() for agent_id in agent_ids}
+            obs = self.observation_space.sample()
```
Comment (gjoliver) [inline rllib/env/multi_agent_env.py:424]: nice!

File: rllib/env/vector_env.py [hunk 3]
```diff
@@ -274,44 +275,70 @@ class VectorEnvWrapper(BaseEnv):
     def __init__(self, vector_env: VectorEnv):
         self.vector_env = vector_env
         self.num_envs = vector_env.num_envs
+
+        # This VectorEnv is multi-agent.
+        self.is_multiagent = False
+        for sub_env in self.vector_env.get_sub_environments():
```
Comment (gjoliver) [inline rllib/env/vector_env.py:281]: should this check live in Avnish's VectorEnv validation module?
we would simply do the following here.
```
self.is_multiagent = any(
    isinstance(e, MultiAgentEnv) for e in self.vector_env.get_sub_environments())
```

File: rllib/env/vector_env.py [hunk 4]
```diff
@@ -321,14 +348,13 @@ def send_actions(self, action_dict: MultiEnvDict) -> None:
...
-        from ray.rllib.env.base_env import _DUMMY_AGENT_ID
-
         assert env_id is None or isinstance(env_id, int)
-        return {
-            env_id
-            if env_id is not None
-            else 0: {_DUMMY_AGENT_ID: self.vector_env.reset_at(env_id)}
-        }
+        if not self.is_multiagent:
+            from ray.rllib.env.base_env import _DUMMY_AGENT_ID
```
Comment (gjoliver) [inline rllib/env/vector_env.py:353]: why import _DUMMY_AGENT_ID again?

File: rllib/env/vector_env.py [hunk 5]
```diff
@@ -321,14 +348,13 @@ def send_actions(self, action_dict: MultiEnvDict) -> None:
...
+
+            return {env_id or 0: {_DUMMY_AGENT_ID: self.vector_env.reset_at(env_id)}}
```
Comment (gjoliver) [inline rllib/env/vector_env.py:355]: better to write with_dummy_agent_id({env_id or 0: self.vector_env.reset_at(env_id)})?

File: rllib/env/vector_env.py [hunk 6]
```diff
@@ -360,16 +386,26 @@ def action_space(self) -> gym.Space:
...
+        else:
+            return {0: self._action_space.sample()}
 
     @override(BaseEnv)
     @PublicAPI
     def observation_space_sample(self, agent_id: list = None) -> MultiEnvDict:
-        del agent_id
-        return {0: {_DUMMY_AGENT_ID: self._observation_space.sample()}}
+        if not self.is_multiagent:
+            return {0: {_DUMMY_AGENT_ID: self._observation_space.sample()}}
```
Comment (gjoliver) [inline rllib/env/vector_env.py:398]: maybe try to use with_dummy_agent_id() here too.

File: rllib/evaluation/sampler.py [hunk 7]
```diff
@@ -834,6 +834,10 @@ def _process_observations(
             episode._add_agent_rewards(rewards[env_id])
 
         # Check episode termination conditions.
+        try:  # TODO
```
Comment (gjoliver) [inline rllib/evaluation/sampler.py:837]: debugging code?

**Review by gjoliver — COMMENTED** [review]
cool, like this PR a lot in general.
just a bunch of nits.

**Comment by matteobettini** [comment]
Hello,

I just wnated to know if Vectorized MultiAgentEnvs are supported yet? 
Would implementing something like ` MockMultiAgentVectorEnv` work now in rllib?

Thanks

**Comment by matteobettini** [comment]
@sven1977 @smorad  will this ever be a thing or we'll never be able to have vectorized multi-agent environmnents?

**Pull request closed without merging**